src/sigchain/Sigchain.ts

```diff
@@ -93,7 +91,7 @@ class Sigchain {
     // Therefore, with no claims in the sigchain, the previous sequence number
     // is set to 0.
     await withF([this.db.transaction()], async ([tran]) => {
-      const sequenceNumber = await tran.get<number | null>([
+      const sequenceNumber = await tran.getForUpdate<number | null>([
```

Comment:
Does this need a `getForUpdate`?

This would be in case you're going to do a `put` later involving the same key, and whether this could occur concurrently.

When we are in a singleton context, it shouldn't be necessary. But I'm debating whether it's better to be safe. But my understanding is that `start` is never called concurrently (and if it is, that's handled by the decorator).

What is the race condition here?

Reply:
The same key is updated updated here so I used the `getForUpdate`. if it doesn't really make a difference here due to the life cycle locking then I'd lean towards keeping it for consistency.

Reply:
Yes, but at the same time, this is only useful during a write skew. Which means 2 overlapping reading a shared set of data then writing puts that violates constraints.

Does this strictly happen here even without lifecycle locking?

Reply:
I should clarify:

1. With SI as we have it now, there is no possibility of consistency anomalies anymore. So technically race conditions cannot occur with proper use of transactions.
2. Write skew is the only anomaly left, which is rare.
3. Counter racing is technically not a race condition, we don't break the consistency of our data model with it due to SI. What happens more accurately is that we end up with lots of repeating transaction conflicts. To avoid this, we use counter locking, the `getForUpdate` is never needed here

Reply:
So an example of a write skew in our system right now is the relationship between vault id and vault name.

If 2 transactions are updating a vault id's vault name. They could both read from the same vault id, and then set different vault names. Since vault id and vault names are meant to be 1 to 1 relationships, this could break consistency without having a transaction conflict.

2 things should be done here:

1. Add a lock to serialise all vault id vault name updates.
2. Change to using `getForUpdate` for the vault id

The 2. step isn't technically necessary because 1. already serialises it. But 2. should be added because it's a second layero f defense and most importantly it hints to the developer that this is a write skew.

Reply:
This has been removed, `getForUpdate` is only used for adding new claims to avoid diverging history.